README.md
# gtmetrix

Node.js module for the GTmetrix API to run and access tests.

[![npm](https://img.shields.io/npm/v/gtmetrix.svg?maxAge=3600)](https://github.com/fvdm/nodejs-gtmetrix/blob/master/CHANGELOG.md)
[![Build Status](https://travis-ci.org/fvdm/nodejs-gtmetrix.svg?branch=master)](https://travis-ci.org/fvdm/nodejs-gtmetrix)
[![Coverage Status](https://coveralls.io/repos/github/fvdm/nodejs-gtmetrix/badge.svg?branch=master)](https://coveralls.io/github/fvdm/nodejs-gtmetrix?branch=master)
[![Dependencies](https://www.bithound.io/github/fvdm/nodejs-gtmetrix/badges/dependencies.svg)](https://www.bithound.io/github/fvdm/nodejs-gtmetrix/develop/dependencies/npm)
[![Code Quality](https://www.bithound.io/github/fvdm/nodejs-gtmetrix/badges/code.svg)](https://www.bithound.io/github/fvdm/nodejs-gtmetrix)
[![Greenkeeper](https://badges.greenkeeper.io/fvdm/nodejs-gtmetrix.svg)](https://greenkeeper.io/)

* [Node.js](https://nodejs.org)
* [GTMetrix](https://gtmetrix.com)
* [API documentation](https://gtmetrix.com/api/)


## Example

```js
const gtmetrix = require ('gtmetrix') ({
  email: 'your@mail.tld',
  apikey: 'abc123'
});

// Run test from London with Google Chrome
const testDetails = {
  url: 'http://example.net/',
  location: 2,
  browser: 3
};

// Poll test every 5 seconds for completion, then log the result
gtmetrix.test.create (testDetails).then (data =>
  gtmetrix.test.get (data.test_id, 5000).then (console.log));
```

_(For readability I left out the error handling)_


## Installation

`npm i gtmetrix`

You need an account at GTmetrix to get an API key.
The API key can be found [here](https://gtmetrix.com/api/#api-details) when you
are logged in.


## Configuration

The setup function takes an _object_ with these settings.

name    | type    | required | default | description
:-------|:--------|:---------|:--------|:--------------------
email   | string  | yes      |         | Your account email
apikey  | string  | yes      |         | Your account API key
timeout | integer | no       | 5000    | Wait timeout in ms


#### Example

```js
const gtmetrix = require ('gtmetrix') ({
  email: 'your@mail.tld',
  apikey: 'abc123',
  timeout: 10000
});
```


## Methods

All methods return promises, but you can also provide a callback function
instead which gets the standard `err` and `data` arguments.

In the examples below I use a mix of callbacks and promises, but each method
can do both. I also left out the error handling for better readability.


#### Errors

message          | description            | properties
:----------------|:-----------------------|:----------------------------------
request failed   | Request cannot be made | `error`
invalid response | Can't process response | `error` `statusCode` `contentType`
API error        | API returned an error  | `error` `statusCode` `contentType`


### test.create
**( params, [callback] )**

Run a test.

argument | type     | required | description
:--------|:---------|:---------|:-----------------
params   | object   | yes      | Test settings
callback | function | no       | `(err, data)` or promise

[API documentation](https://gtmetrix.com/api/#api-test-start)


```js
// Run test from London with Google Chrome
const test = {
  url: 'http://example.net/',
  location: 2,
  browser: 3
};

gtmetrix.test.create (test, console.log);
```

```js
{ credits_left: 68,
  test_id: 'Ao0AYQbz',
  poll_state_url: 'https://gtmetrix.com/api/0.1/test/Ao0AYQbz' }
```


### test.get
**( testId, [resource], [polling], [callback] )**

Get details about a test or one of its resources.

When you specify a binary resource, i.e. `screenshot`,
the callback `data` will be a _Buffer_ instance, so you can
post-process the binary data however you like. See example below.

argument | type     | required | description
:--------|:---------|:---------|:-----------------------------
testId   | string   | yes      | Test `id` to look up
resource | string   | no       | Retrieve a test resource
polling  | number   | no       | Retry until completion, in ms
callback | function | no       | `(err, data)` or promise

[API documentation](https://gtmetrix.com/api/#api-test-state)


#### Test details

Get what is currently available, without waiting for completion.

```js
gtmetrix.test.get ('Ao0AYQbz', console.log);
```

```js
{ resources: {}, error: '', results: {}, state: 'started' }
```


Wait for completion and then get the details.

```js
// Get test result when it is complete, retry every 5 seconds (5000 ms)
gtmetrix.test.get ('Ao0AYQbz', 5000, console.log);
```

```js
{ resources:
   { report_pdf: 'https://gtmetrix.com/api/0.1/test/Ao0AYQbz/report-pdf',
     pagespeed: 'https://gtmetrix.com/api/0.1/test/Ao0AYQbz/pagespeed',
     har: 'https://gtmetrix.com/api/0.1/test/Ao0AYQbz/har',
     pagespeed_files: 'https://gtmetrix.com/api/0.1/test/Ao0AYQbz/pagespeed-files',
     report_pdf_full: 'https://gtmetrix.com/api/0.1/test/Ao0AYQbz/report-pdf?full=1',
     yslow: 'https://gtmetrix.com/api/0.1/test/Ao0AYQbz/yslow',
     screenshot: 'https://gtmetrix.com/api/0.1/test/Ao0AYQbz/screenshot' },
  error: '',
  results:
   { onload_time: 185,
     first_contentful_paint_time: 221,
     page_elements: 2,
     report_url: 'https://gtmetrix.com/reports/example.net/Ao0AYQbz',
     redirect_duration: 0,
     first_paint_time: 221,
     dom_content_loaded_duration: null,
     dom_content_loaded_time: 184,
     dom_interactive_time: 183,
     page_bytes: 1911,
     page_load_time: 185,
     html_bytes: 277,
     fully_loaded_time: 307,
     html_load_time: 145,
     rum_speed_index: 221,
     yslow_score: 99,
     pagespeed_score: 99,
     backend_duration: 68,
     onload_duration: 0,
     connect_duration: 77 },
  state: 'completed' }
```


#### Retrieve screenshot

Retry every 5000 ms until it's ready.

```js
const fs = require ('fs');

gtmetrix.test.get ('Ao0AYQbz', 'screenshot', 5000).then (data =>
  fs.writeFile (__dirname + '/screenshot.jpg', data, console.log));
```


##### Resources

resource        | binary | content   | description
:---------------|:-------|:----------|:---------------------------------------
filmstrip       | yes    | JPEG      | Page loading filmstrip (requires video)
har             | no     | JS object | HTTP Archive
pagespeed       | no     | JS object | Pagespeed report
pagespeed-files | yes    | ZIP       | Pagespeed optimized files
report-pdf      | yes    | PDF       | Test summary
report-pdf-full | yes    | PDF       | Full test report
screenshot      | yes    | JPEG      | Screenshot image
video           | yes    | MP4       | Page loading video
yslow           | no     | JS object | YSlow report


### locations.list
**( [callback] )**

Get a list of available test locations.

argument | type     | required | description
:--------|:---------|:---------|:------------------------
callback | function | no       | `(err, data)` or promise

[API documentation](https://gtmetrix.com/api/#api-locations)


```js
gtmetrix.locations.list (console.log);
```

```js
[ { name: 'Vancouver, Canada',
    default: true,
    id: '1',
    browsers: [ 1, 3 ] } ]
```


### browsers.list
**( [callback] )**

Get a list of available test browsers.

argument | type     | required | description
:--------|:---------|:---------|:------------------------
callback | function | no       | `(err, data)` or promise

[API documentation](https://gtmetrix.com/api/#api-browsers)


```js
gtmetrix.browsers.list (console.log);
```

```js
[ { features:
     { dns: true,
       cookies: true,
       adblock: true,
       http_auth: true,
       video: true,
       user_agent: true,
       throttle: true,
       filtering: true,
       resolution: true },
    browser: 'firefox',
    name: 'Firefox (Desktop)',
    platform: 'desktop',
    id: 1,
    device: '' } ]
```


### browsers.get
**( browserId, [callback] )**

Get details about a test browser.

argument  | type     | required | description
:---------|:---------|:---------|:-----------------------
browserId | number   | yes      | Browser `id` to look up
callback  | function | no       | `(err, data)` or promise

[API documentation](https://gtmetrix.com/api/#api-browsers-details)


```js
gtmetrix.browsers.get (3, console.log);
```

```js
{ features:
   { dns: true,
     cookies: true,
     adblock: true,
     http_auth: true,
     video: true,
     user_agent: true,
     throttle: true,
     filtering: true,
     resolution: true },
  browser: 'chrome',
  name: 'Chrome (Desktop)',
  platform: 'desktop',
  id: 3,
  device: '' }
```


### account.status
**( [callback] )**

Information about your account.

argument | type     | required | description
:--------|:---------|:---------|:------------------------
callback | function | no       | `(err, data)` or promise

[API documentation](https://gtmetrix.com/api/#api-status)


```js
gtmetrix.account.status (console.log);
```

```js
{ api_refill: 1234567890, api_credits: 68 }
```


## Unlicense

This is free and unencumbered software released into the public domain.

Anyone is free to copy, modify, publish, use, compile, sell, or
distribute this software, either in source code form or as a compiled
binary, for any purpose, commercial or non-commercial, and by any
means.

In jurisdictions that recognize copyright laws, the author or authors
of this software dedicate any and all copyright interest in the
software to the public domain. We make this dedication for the benefit
of the public at large and to the detriment of our heirs and
successors. We intend this dedication to be an overt act of
relinquishment in perpetuity of all present and future rights to this
software under copyright law.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

For more information, please refer to <https://unlicense.org>


## Author

[Franklin van de Meent](https://frankl.in)

[![Buy me a coffee](https://frankl.in/u/kofi/kofi-readme.png)](https://ko-fi.com/franklin)


CHANGELOG.md
### 1.3.0 (2018-03-06)

##### Documentation Changes

* **readme:**
  *  Add output examples ([a6f65647](https://github.com/fvdm/nodejs-gtmetrix/commit/a6f656472070c3dc21d340d7c2f3478b27cd9bf1))
  *  Callback arg is no longer required ([b20aa309](https://github.com/fvdm/nodejs-gtmetrix/commit/b20aa30925fa0ec0c8f6c9cdb4e94a73f32516bb))
  *  Minor edits ([0a1af5f0](https://github.com/fvdm/nodejs-gtmetrix/commit/0a1af5f0586398a39c800602f2f7567e77ae4a0c))
  *  Better parsable syntax ([3ace947d](https://github.com/fvdm/nodejs-gtmetrix/commit/3ace947d10aead442b25c09da0f37ef01fa98248))
  *  Removed result from example ([6e858c13](https://github.com/fvdm/nodejs-gtmetrix/commit/6e858c13ea5939d540bd3a0b7462e2b003d23f3f))
  *  Full code example ([545b7021](https://github.com/fvdm/nodejs-gtmetrix/commit/545b7021f86c7f9173a75af77414e252b6339b6d))

##### New Features

* **interface:**  Promisified all methods ([54a74b59](https://github.com/fvdm/nodejs-gtmetrix/commit/54a74b59052c514dbf4b008f1eefae2f2ac8e1f3))

##### Refactors

* **interface:**  Switched to es6-promisify ([4dfae869](https://github.com/fvdm/nodejs-gtmetrix/commit/4dfae869d147977959cbd3690529fa3ba6c58b1c))
* **package:**  Minimum node version 8.0.0 ([ad64bb9a](https://github.com/fvdm/nodejs-gtmetrix/commit/ad64bb9a6744d9b4447096678628cf2a7ac2881d))

##### Code Style Changes

* **syntax:**  Only set doRequest of httpreq ([f24606f6](https://github.com/fvdm/nodejs-gtmetrix/commit/f24606f64ba1d954ca6115e66a463394d130e4b3))

#### 1.2.7 (2018-03-02)

##### Bug Fixes

* **config:**  parseInt timeout and defaults ([db336a14](https://github.com/fvdm/nodejs-gtmetrix/commit/db336a142f4b582da931f45fb5f96ab6346138b0))

#### 1.2.6 (2018-03-01)

##### Documentation Changes

* **badges:**  Moved Greenkeeper to the badges ([e07b9f99](https://github.com/fvdm/nodejs-gtmetrix/commit/e07b9f99dfae648f47e1d10337e104572b88de4d))

##### Bug Fixes

* **account:**  Account.status typo ([30619c8a](https://github.com/fvdm/nodejs-gtmetrix/commit/30619c8ac3e48b232fa4971c76c05e5c16704bfb))
* **response:**  Missing references ([0930b8a8](https://github.com/fvdm/nodejs-gtmetrix/commit/0930b8a8aeb739bdd9e3bc7d961b62906904c046))
* **pollingCallback:**
  *  Stop polling when error ([af3ccb0d](https://github.com/fvdm/nodejs-gtmetrix/commit/af3ccb0d12a01d063a2633ad6acfc8149ecbcc37))
  *  Syntax typo ([bc1434c6](https://github.com/fvdm/nodejs-gtmetrix/commit/bc1434c627b5a971ba7fc034db29679c6b2f987c))
  *  Enforce type string on error ([#14](https://github.com/fvdm/nodejs-gtmetrix/pull/14)) ([678bd227](https://github.com/fvdm/nodejs-gtmetrix/commit/678bd227ae7aefa82ac76bc86f27b12098417083))
* **polling:**  Check err.error before match ([#14](https://github.com/fvdm/nodejs-gtmetrix/pull/14)) ([6fe16d0c](https://github.com/fvdm/nodejs-gtmetrix/commit/6fe16d0ca13b6f110a98862361f32b1d66f5eb59))

##### Refactors

* **main:**  Cleaner interface statements ([91e06140](https://github.com/fvdm/nodejs-gtmetrix/commit/91e06140373d590524c15db0425a3f093055afdb))
* **testGet:**  Reduced complexity ([6e1b2738](https://github.com/fvdm/nodejs-gtmetrix/commit/6e1b2738a268ba94d4788cfbb3c4a8f4b262de84))
* **pollingCallback:**
  *  Reduced complexity ([c4bf94b2](https://github.com/fvdm/nodejs-gtmetrix/commit/c4bf94b22ab8b7408e6216c2264d341863d10a70))
  *  Rewrite response handling ([ad2bc570](https://github.com/fvdm/nodejs-gtmetrix/commit/ad2bc570297387a91f5dcb0dc049641483fca979))
  *  Clean up redundancy ([9de29973](https://github.com/fvdm/nodejs-gtmetrix/commit/9de29973915b9e1fb30ca76ad335f1022628115e))

##### Tests

* **main:**
  *  Fixed bad ref ([bafaf5c2](https://github.com/fvdm/nodejs-gtmetrix/commit/bafaf5c250a1c1c47df630628096dc449d589e65))
  *  Fixed bad test var ([00059abe](https://github.com/fvdm/nodejs-gtmetrix/commit/00059abe91be356b0089035d06c2ada31dc57214))
  *  Add Interface test ([34209410](https://github.com/fvdm/nodejs-gtmetrix/commit/342094104466db7984343f3a2aef724c3ff141a2))
* **package:**  Change dev deps to dotest ([cd347553](https://github.com/fvdm/nodejs-gtmetrix/commit/cd347553e8ef185c66b0c8a71423973a06f66790))
* **config:**  Update Travis CI node versions ([32073c21](https://github.com/fvdm/nodejs-gtmetrix/commit/32073c21eb1b3e10f735e471f025c7711a63a8ad))

#### 1.2.5 (2017-09-09)

##### Chores

* **repo:** Removed package-lock.json ([fef5aa2a](https://github.com/fvdm/nodejs-gtmetrix/commit/fef5aa2ac202e32e2948ed562365a4f5cdd1c271))

#### 1.2.4 (2017-7-8)

##### Chores

* **repo:** Added package-lock.json ([194a2651](https://github.com/fvdm/nodejs-gtmetrix/commit/194a2651bf8972cbc85f2fc0cab54d998a160f20))
* **package:**
  * Clean up description and keywords ([7bdcf6da](https://github.com/fvdm/nodejs-gtmetrix/commit/7bdcf6dadf49867e46f1e5ec80cd6f76b136dcfe))
  * Updated dependencies ([c9539360](https://github.com/fvdm/nodejs-gtmetrix/commit/c95393604fe0a5e4b855b7ec7ce18158e1d0d59e))
  * update eslint to version 3.19.0 (#13) ([840d95d9](https://github.com/fvdm/nodejs-gtmetrix/commit/840d95d9968e8f6841c9fce8d5f0ef3dc173f9c6))
  * update coveralls to version 2.13.0 (#12) ([e58acb3a](https://github.com/fvdm/nodejs-gtmetrix/commit/e58acb3a275750c1f39e9aaae4c93fd2a01ac0df))
  * Update dotest dev dep ([ebf48a19](https://github.com/fvdm/nodejs-gtmetrix/commit/ebf48a19d26892103ba105f6d28ec2c228c7dd2d))
  * Update dev deps ([52a28531](https://github.com/fvdm/nodejs-gtmetrix/commit/52a28531c801f64c2dc2b42188328da4bcf9d937))

##### Documentation Changes

* **readme:**
  * Minor clean up ([e78ae317](https://github.com/fvdm/nodejs-gtmetrix/commit/e78ae31716481a28911c86ba365775ed84c1d9fd))
  * Add coffee button to Author ([e9c4194a](https://github.com/fvdm/nodejs-gtmetrix/commit/e9c4194ae4496809c8dc2cd4b65aa59ca89739df))

##### Bug Fixes

* **testGet:** Get resource with polling failed ([bb909295](https://github.com/fvdm/nodejs-gtmetrix/commit/bb909295afe122360992d5c7e1d1c9b4a0fc90af))

##### Refactors


test.js
/*
Name:           gtmetrix - test.js
Author:         Franklin van de Meent (https://frankl.in)
Source & docs:  https://github.com/fvdm/nodejs-gtmetrix
Feedback:       https://github.com/fvdm/nodejs-gtmetrix/issues
License:        Unlicense / Public Domain (see UNLICENSE file)
                (https://github.com/fvdm/nodejs-gtmetrix/raw/develop/LICENSE)
*/

var path = require ('path');
var dotest = require ('dotest');
var app = require (path.join (__dirname, path.sep));


// Setup
// $ GTMETRIX_EMAIL= GTMETRIX_APIKEY= npm test
var email = process.env.GTMETRIX_EMAIL;
var apikey = process.env.GTMETRIX_APIKEY;
var timeout = String (process.env.GTMETRIX_TIMEOUT);
var location = process.env.GTMETRIX_LOCATION || 2;
var browser = process.env.GTMETRIX_BROWSER || 3;

var gtmetrix = app ({
  email: email,
  apikey: apikey,
  timeout: timeout
});

var cache = {
  url: 'http://example.net/',
  location,
  browser
};


dotest.add ('Interface', test => {
  const tst = gtmetrix && gtmetrix.test;
  const locations = gtmetrix && gtmetrix.locations;
  const browsers = gtmetrix && gtmetrix.browsers;
  const account = gtmetrix && gtmetrix.account;

  test()
    .isFunction ('fail', 'export', app)
    .isObject ('fail', 'module', gtmetrix)

    .isObject ('fail', '.test', tst)
    .isFunction ('fail', '.test.create', tst && tst.create)
    .isFunction ('fail', '.test.get', tst && tst.get)

    .isObject ('fail', '.locations', locations)
    .isFunction ('fail', '.locations.list', locations && locations.list)

    .isObject ('fail', '.browsers', browsers)
    .isFunction ('fail', '.browsers.list', browsers && browsers.list)
    .isFunction ('fail', '.browsers.get', browsers && browsers.get)

    .isObject ('fail', '.account', account)
    .isFunction ('fail', '.account.status', account && account.status)
    .done();
});


/**
 * Promise handling
 */

dotest.add ('Promise - account.status', test => {
  gtmetrix.account.status ()
    .catch (err => test (err).done())
    .then (data => test()
      .isObject ('fail', 'data', data)
      .done()
    );
});


dotest.add ('Promise - Error: API error - without polling', test => {
  gtmetrix.test.get ('0')
    .catch (err => {
      test()
        .isError ('fail', 'err', err)
        .isExactly ('fail', 'err.message', err && err.message, 'API error');
    })
    .then (data => {
      test()
        .isUndefined ('fail', 'data', data)
        .done();
    });
});


/**
 * Callback handling
 */

dotest.add ('Callback - account.status', test => {
  gtmetrix.account.status ((err, data) => {
    test (err)
      .isObject ('fail', 'data', data)
      .done ();
  });
});


dotest.add ('Callback - browsers.list', test => {
  gtmetrix.browsers.list ((err, data) => {
    test (err)
      .isArray ('fail', 'data', data)
      .isNotEmpty ('fail', 'data', data)
      .isObject ('warn', 'data[0]', data && data [0])
      .done ();
  });
});


dotest.add ('Callback - browsers.get', test => {
  gtmetrix.browsers.get (3, (err, data) => {
    test (err)
      .isObject ('fail', 'data', data)
      .isExactly ('warn', 'data.id', data && data.id, 3)
      .done ();
  });
});


dotest.add ('Callback - locations.list', test => {
  gtmetrix.locations.list ((err, data) => {
    test (err)
      .isArray ('fail', 'data', data)
      .isNotEmpty ('fail', 'data', data)
      .isObject ('warn', 'data[0]', data && data [0])
      .done ();
  });
});


dotest.add ('Callback - test.create', test => {
  gtmetrix.test.create (cache, (err, data) => {
    cache.test = data;
    test (err)
      .isObject ('fail', 'data', data)
      .done ();
  });
});


dotest.add ('Callback - test.get - without polling', test => {
  gtmetrix.test.get (cache.test.test_id, (err, data) => {
    test (err)
      .isObject ('fail', 'data', data)
      .isString ('warn', 'data.state', data && data.state)
      .done ();
  });
});


dotest.add ('Callback - test.get - with polling', test => {
  gtmetrix.test.get (cache.test.test_id, 5000, (err, data) => {
    test (err)
      .isObject ('fail', 'data', data)
      .isExactly ('fail', 'data.state', data && data.state, 'completed')
      .done ();
  });
});


dotest.add ('Callback - test.get resource - binary with polling', test => {
  gtmetrix.test.get (cache.test.test_id, 'report-pdf-full', 5000, (err, data) => {
    test (err)
      .isObject ('fail', 'data', data)
      .done ();
  });
});


dotest.add ('Callback - test.get resource - non-binary with polling', test => {
  gtmetrix.test.get (cache.test.test_id, 'yslow', 5000, (err, data) => {
    test (err)
      .isObject ('fail', 'data', data)
      .isExactly ('fail', 'data.u', data && data.u, cache.url)
      .done ();
  });
});


dotest.add ('Callback - Error: API error - without polling', test => {
  gtmetrix.test.get ('0', (err, data) => {
    test ()
      .isError ('fail', 'err', err)
      .isExactly ('fail', 'err.message', err && err.message, 'API error')
      .isUndefined ('fail', 'data', data)
      .done ();
  });
});


dotest.add ('Callback - Error: API error - resource with default polling', test => {
  gtmetrix.test.get ('0', 'yslow', true, (err, data) => {
    test ()
      .isError ('fail', 'err', err)
      .isExactly ('fail', 'err.message', err && err.message, 'API error')
      .isUndefined ('fail', 'data', data)
      .done ();
  });
});


dotest.add ('Callback - Error: request failed (timeout)', test => {
  var tmp = app ({
    email: email,
    apikey: apikey,
    timeout: 1
  });

  tmp.account.status ((err, data) => {
    var error = err && err.error;

    test ()
      .isError ('fail', 'err', err)
      .isExactly ('fail', 'err.message', err && err.message, 'request failed')
      .isError ('fail', 'err.error', error)
      .isExactly ('warn', 'err.error.code', error && error.code, 'TIMEOUT')
      .isUndefined ('fail', 'data', data)
      .done ();
  });
});


// Start the tests
dotest.run (1000);


index.js
const { doRequest } = require ('httpreq');
const { promisify } = require ('es6-promisify');


// Default config
const defaults = {
  email: null,
  apikey: null,
  timeout: 5000
};

let pkg = {
  test: {
    get: null
  },
  locations: {},
  browsers: {},
  account: {}
};

let config = {};


/**
 * Get info about resource type
 *
 * @param   {string}  name  The resource name
 * @return  {object}        Resource info object
 */

function resourceType (name) {
  let info = {
    binary: false,
    path: name.replace ('_', '-')
  };

  if (info.path.match (/^(filmstrip|pagespeed-files|report-pdf(-full)?|screenshot|video)$/)) {
    info.binary = true;
  }

  if (info.path === 'report-pdf-full') {
    info.path = 'report-pdf?full=1';
  }

  return info;
}


/**
 * Make an error
 *
 * @return  {Error}
 *
 * @param   {string}       msg   Error.message
 * @param   {mixed}        err   Error.error
 * @param   {number|null}  code  Error.statusCode
 * @param   {string|null}  type  Error.contentType
 */

function doError (msg, err, code, type) {
  let error = new Error (msg);

  error.error = err;
  error.statusCode = code;
  error.contentType = type;
  return error;
}


/**
 * Process API response
 *
 * @callback  callback
 * @return    {void}
 *
 * @param     {object}      options   httpreq options
 * @param     {Error|null}  err       httpreq Error
 * @param     {object}      res       httpreq response
 * @param     {function}    callback  `(err, data)`
 */

function apiResponse (options, err, res, callback) {
  let type;
  let size;
  let code;
  let data;
  let error = null;

  if (err) {
    error = doError ('request failed', err, null, null);
    callback (error);
    return;
  }

  type = String (res.headers ['content-type']) .split (';') [0];
  size = String (res.headers ['content-length']);
  code = res.statusCode;
  data = res.body;

  // Received data, expecting binary
  if (size && options.binary && type.match (/\/(pdf|jpeg|tar)$/)) {
    callback (null, data);
    return;
  }

  // Received something else
  try {
    data = JSON.parse (data);

    if (data.error) {
      error = doError ('API error', data.error, code, type);
      callback (error);
      return;
    }
  } catch (e) {
    error = doError ('invalid response', e, code, type);
    callback (error);
    return;
  }

  // It's real data
  callback (null, data);
}


/**
 * Send API request
 *
 * @callback  callback
 * @return    {void}
 *
 * @param     {object}    props
 * @param     {boolean}   [props.binary=false]  Expect binary response
 * @param     {string}    props.method          HTTP method
 * @param     {object}    [props.params]        Method parameters
 * @param     {string}    props.path            Method path
 * @param     {function}  callback              `(err, data)`
 */

function apiRequest (props, callback) {
  const options = {
    url: 'https://gtmetrix.com/api/0.1/' + props.path,
    parameters: props.params || null,
    method: props.method,
    headers: {
      'User-Agent': 'gtmetrix.js (https://www.npmjs.com/package/gtmetrix)'
    },
    timeout: parseInt (config.timeout, 10) || defaults.timeout,
    auth: config.email + ':' + config.apikey,
    binary: props.binary || false
  };

  doRequest (options, (err, res) => {
    apiResponse (options, err, res, callback);
  });
}


/**
 * Create test
 *
 * @callback  callback
 * @return    {Promise<object>}
 *
 * @param     {object}    params
 * @param     {function}  [callback]  `(err, data)`
 */

pkg.test.create = promisify ((params, callback) => {
  const props = {
    method: 'POST',
    path: 'test',
    params: params
  };

  apiRequest (props, callback);
});


/**
 * Process callback when polling
 *
 * @callback  callback
 * @return    {boolean}               Stop polling? true = yes
 *
 * @param     {object}      props     Request properties
 * @param     {Error|null}  err       Response error
 * @param     {object}      data      Response data
 * @param     {function}    callback  `(err, data)`
 */

function pollingCallback (props, err, data, callback) {
  // API error saying we need to wait
  if (err && String (err.error).match (/Data not yet available/)) {
    return false;
  }

  // Another API error
  if (err) {
    callback (err);
    return true;
  }

  // No API error, binary expected = ok complete
  if (props.binary) {
    callback (null, data);
    return true;
  }

  // No error, non-binary, not running = ok complete
  if (!String (data.state).match (/^(started|queued)$/)) {
    callback (null, data);
    return true;
  }

  // else keep polling
  return false;
}


/**
 * Process test response
 *
 * @callback  callback
 * @return    {void}
 *
 * @param     {object}      params
 * @param     {bool}        params.polling   Keep polling for updates
 * @param     {string}      params.testId    Test ID
 * @param     {string}      params.resource  Test resource
 * @param     {object}      params.props     Request params
 * @param     {Error|null}  err              Response error
 * @param     {mixed}       data             Response data
 * @param     {function}    callback         `(err, data)`
 */

function testResponse (params, err, data, callback) {
  let retryInterval;
  let complete;

  if (err && !params.polling) {
    callback (err);
    return;
  }

  if (!params.polling) {
    callback (null, data);
    return;
  }

  if (params.polling === true) {
    params.polling = 5000;
  }

  if (typeof params.polling === 'number') {
    complete = pollingCallback (params.props, err, data, callback);

    if (complete) {
      return;
    }

    // Test is still running
    retryInterval = setInterval (() => {
      pkg.test.get (params.testId, params.resource, (pErr, pData) => {
        const pComplete = pollingCallback (params.props, pErr, pData, callback);

        if (pComplete) {
          clearInterval (retryInterval);
        }
      });
    }, params.polling);
  }
}


/**
 * Get test result
 *
 * @callback  callback
 * @return    {Promise<object>}
 *
 * @param     {string}    testId      Test ID
 * @param     {string}    [resource]  Resource to get, i.e. `screenshot`
 * @param     {number}    [polling]   Poll state until completion, in ms
 * @param     {function}  [callback]  `(err, data)`
 */

pkg.test.get = promisify ((testId, resource, polling, callback) => {
  let resourceInfo = {};
  let params = {
    testId,
    resource,
    polling,
    props: {
      method: 'GET',
      path: 'test/' + testId
    }
  };

  if (typeof polling === 'function') {
    callback = polling;
    params.polling = null;
  }

  switch (typeof resource) {
    case 'function':
      callback = resource;
      params.resource = null;
      params.polling = null;
      break;

    case 'number':
      params.polling = resource;
      params.resource = null;
      break;

    case 'string':
      resourceInfo = resourceType (resource);
      params.props.path += '/' + resourceInfo.path;
      params.props.binary = resourceInfo.binary;
      break;

    default:
      break;
  }

  apiRequest (params.props, (err, data) => {
    testResponse (params, err, data, callback);
  });
});


/**
 * List locations
 *
 * @callback  callback
 * @return    {Promise<object>}
 *
 * @param     {function}  [callback]  `(err, data)`
 */

pkg.locations.list = promisify ((callback) => {
  const props = {
    method: 'GET',
    path: 'locations'
  };

  apiRequest (props, callback);
});


/**
 * List browsers
 *
 * @callback  callback
 * @return    {Promise<object>}
 *
 * @param     {function}  [callback]  `(err, data)`
 */

pkg.browsers.list = promisify ((callback) => {
  const props = {
    method: 'GET',
    path: 'browsers'
  };

  apiRequest (props, callback);
});


/**
 * Get browser
 *
 * @callback  callback
 * @return    {Promise<object>}
 *
 * @param     {string}    browserId
 * @param     {function}  [callback]  `(err, data)`
 */

pkg.browsers.get = promisify ((browserId, callback) => {
  const props = {
    method: 'GET',
    path: 'browsers/' + browserId
  };

  apiRequest (props, callback);
});


/**
 * Get account status
 *
 * @callback  callback
 * @return    {Promise<object>}
 *
 * @param     {function}  [callback]  `(err, data)`
 */

pkg.account.status = promisify ((callback) => {
  const props = {
    method: 'GET',
    path: 'status'
  };

  apiRequest (props, callback);
});


/**
 * Module interface
 *
 * @return  {object}                        Methods
 *
 * @param   {object}  props
 * @param   {string}  props.email           API email
 * @param   {string}  props.apikey          API key
 * @param   {number}  [props.timeout=5000]  Request timeut in ms
 */

module.exports = (props) => {
  let key;

  for (key in props) {
    config [key] = props [key];
  }

  return pkg;
};
